### ISSUE TITLE
Support --raw-output0

### ISSUE THREAD

**Issue by pkoppstein** [01mf02/jaq#101]
The 1.0 beta announcement mentions support for the --nul-output command-line option.

Please note that jq 1.7rc1 has removed that option, and mentions instead:
```
      --raw-output0         implies -r and output NUL after each output;
```
It is generally expected that jq 1.7 will be released quite soon, and that the above remarks about jq 1.7rc1 will apply to 1.7 as well.

**Comment by baod-rate** [comment]
Thanks for the heads up. This was introduced in https://github.com/jqlang/jq/pull/2684. Worth noting that `--raw-output0` bails when any outputs contain a NUL:

https://github.com/jqlang/jq/blob/a6920601299913ec5f9c9a8883517ec10dd71ffe/docs/content/manual/manual.yml#L150-L154

On top of renaming `--nul-output` to `--raw-output0`, we should add that check to remain consistent with jq.

note that the behavior of `-j`/`--join-output` was already inconsistent with jq (in jaq it does not imply `-r`). I matched that in `--nul-output` for internal consistency.

**Comment by 01mf02** [comment]
For now, I have reverted your previous PR, @baod-rate, because of the change in jq.
I think that it would be great to have some option to set the "output terminator string"; let's call it `--output-sep` for the sake of discussion.
Then the default would be `--output-sep "\n"`, `--join` would be equivalent to `--output-sep ""`, and the `--raw-output0` option would imply `--output-sep "\0"`. On Windows, `--output-sep "\r\n"` could be useful.
Prohibiting certain characters in the output as foreseen by `--raw-output0` seems difficult, though.

**Comment by pjungkamp** [comment]
@01mf02 What are your thoughts on adding a `Format::Raw0` to the polyglot formats in `jaq-fmts`?
The `--raw-output0` could be translated into a `--to raw0`. Prohibiting the null character in a string could now be handled in `jaq_fmts::write::write`.

I've only taken a quick dive into the format-related code, so forgive me if I'm wrong, but it seems to me that the following should suffice:

1. Add a `Raw0` variant to the `Format` enum that's parsed from `raw0` in `jaq-fmts/src/lib.rs`.
2. Fix all match statements in `jaq-fmts/src/read/formats.rs` and `jaq-fmts/src/write/formats.rs`.
3. Add `"raw-output0" => self.to = Some(Format::Raw0),` to the `long` argument matching in `jaq/src/cli.rs`.

Should there be a `fromraw0`/`toraw0` function? 

---

There's already a handy helper for splitting a `BufRead` on a specific byte in [`bstr::io::BufReadExt::byte_records`](https://docs.rs/bstr/1.12.1/bstr/io/trait.BufReadExt.html#method.byte_records).

**Comment by 01mf02** [comment]
> [@01mf02](https://github.com/01mf02) What are your thoughts on adding a `Format::Raw0` to the polyglot formats in `jaq-fmts`?

@pjungkamp, in principle, I'm OK with adding functionality to enable `--raw-output0`. There are some open questions, though:

- All formats so far can be used with `--from` and `--to`. If we allow `--to raw0`, then I think it would also be good to allow `--from raw0`. However, what should that do? Disallow `\0` in input? Split on `\0` in input? I tend towards splitting on `\0`. However, it would be good to be on the same page with upstream jq, so that a potential future `--raw-input0` in jq does the same as jaq.
- Is `raw0` some kind of special case of `raw`? If so, it might make sense to have an output format `Raw(char)`, where `char` is the terminator.
- Should we really reproduce the behaviour that the null character should be prohibited in output? It's somewhat inconsistent with `--raw-output`, given that `--raw-output` allows `\n` in its output.

> Should there be a `fromraw0`/`toraw0` function?

I think not, because there are no `fromraw`/`toraw` functions.

I would say, if you are motivated to implement this, go ahead, and let's look at the implications together. Your overall plan of implementation looks sound to me.

**Comment by pjungkamp** [comment]
I've just opened a PR with a simple implementation of `raw0`. I don't think that `raw0` is a special case of `raw`. Your current implementation has special handling of `\r` characters when parsing the lines of the `raw` format using [`bstr::io::BufReadExt::byte_lines`](https://docs.rs/bstr/1.12.1/bstr/io/trait.BufReadExt.html#method.byte_lines).

One could argue that the `Raw0` format is a special case of a format delimited by any other character, but I think that null-terminated string parsing is of special importance for safe shell scripting. There's a reason that many UNIX cli tools have switches for zero-terminated string handling (e.g. `find -print0`, `xargs -0`, `ls --zero`, `grep -Z`, `sort -z`, `read -d ''` in bash). I don't really see a use-case for a more general any-character-delimited string format.

**Comment by 01mf02** [comment]
> One could argue that the `Raw0` format is a special case of a format delimited by any other character, but I think that null-terminated string parsing is of special importance for safe shell scripting. There's a reason that many UNIX cli tools have switches for zero-terminated string handling (e.g. `find -print0`, `xargs -0`, `ls --zero`, `grep -Z`, `sort -z`, `read -d ''` in bash). I don't really see a use-case for a more general any-character-delimited string format.

True. I was mostly trying to suggest `Raw(char)` because I suspected that it would reduce the implementation size, but looking at your implementation, the `Raw(char)` approach does not seem to bring particular benefits.
